### PR TITLE
Fix camera play promise + clean up

### PR DESCRIPTION
**Pull request by CedricGuillemet**
- fix issue introduced with this PR : https://github.com/BabylonJS/Babylon.js/pull/10609
- Removed duplicated definition of TextureData
- Use 4 textures buffer for getting webcam image while rendering with another

### CODE REVIEW

File: Plugins/NativeCamera/Source/NativeVideo.cpp [hunk 1]
```diff
@@ -140,14 +140,17 @@ namespace Babylon::Plugins
...
+    Napi::Value NativeVideo::Play(const Napi::CallbackInfo& info)
     {
         if (!m_IsPlaying)
         {
             m_IsPlaying = true;
             NativeCameraImpl->Open(m_width, m_height, m_frontCamera);
             RaiseEvent("playing");
         }
+
+        auto deferred = Napi::Promise::Deferred::New(info.Env());
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply Plugins/NativeCamera/Source/NativeVideo.cpp:154]: Fixed.

File: Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm [hunk 2]
```diff
@@ -39,13 +45,23 @@ - (id)init:(std::shared_ptr<Babylon::Plugins::Camera::Impl::ImplData>)implData;
...
+                    CFRelease(textureRef);
+                }
+            }
         }
-        
+
         CameraTextureDelegate* cameraTextureDelegate{};
         AVCaptureSession* avCaptureSession{};
         CVMetalTextureCacheRef textureCache{};
+        CVMetalTextureRef texture[TextureBufferCount]{};
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm]: Why not.

File: Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm [hunk 3]
```diff
@@ -17,6 +17,12 @@
 #include "NativeCameraImpl.h"
 #include <napi/napi.h>
 
+namespace
+{
+    // TODO: is there a way to sync texture use with bgfx?
+    static constexpr unsigned int TextureBufferCount = 4;
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm]: I've added a bit more explanation.

File: Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm [hunk 4]
```diff
@@ -39,13 +40,27 @@ - (id)init:(std::shared_ptr<Babylon::Plugins::Camera::Impl::ImplData>)implData;
...
+            }
         }
-        
+
         CameraTextureDelegate* cameraTextureDelegate{};
         AVCaptureSession* avCaptureSession{};
         CVMetalTextureCacheRef textureCache{};
+        /*
+         AVfoundation might update a texture while it's still being processed by the rendering. If it happens, slowdowns or no update at all will occur.
+         To fix this issue, a texture array is used with enough slots. 1 texture is used for rendering while another is updated.
```
Comment (bghgary) [inline Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm:57]: This explanation sounds like there should be 2 textures. Why 4?

File: Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm [hunk 5]
```diff
@@ -154,27 +169,29 @@ - (id)init:(std::shared_ptr<Babylon::Plugins::Camera::Impl::ImplData>)implData {
...
+    const auto textureCycleCount{implData->texture.size()};
+    int back = (implData->textureFrontIndex+1) % textureCycleCount;
+    int front = implData->textureFrontIndex % textureCycleCount;
+    CVReturn status = CVMetalTextureCacheCreateTextureFromImage(NULL, implData->textureCache, pixelBuffer, NULL, pixelFormat, width, height, 0, &implData->texture[front]);
+    if (status == kCVReturnSuccess)
     {
-        dispatch_async(dispatch_get_main_queue(), ^{
-            implData->textureBGRA = textureBGRA;
-        });
+        if(implData->texture[back])
```
Comment (bghgary) [inline Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm:185]: nit
```suggestion
        if (implData->texture[back])
```

File: Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm [hunk 6]
```diff
@@ -154,27 +169,29 @@ - (id)init:(std::shared_ptr<Babylon::Plugins::Camera::Impl::ImplData>)implData {
...
-    {
-        textureBGRA = CVMetalTextureGetTexture(texture);
-        CFRelease(texture);
-    }
 
-    if(textureBGRA != nil)
+    const auto textureCycleCount{implData->texture.size()};
+    int back = (implData->textureFrontIndex+1) % textureCycleCount;
+    int front = implData->textureFrontIndex % textureCycleCount;
+    CVReturn status = CVMetalTextureCacheCreateTextureFromImage(NULL, implData->textureCache, pixelBuffer, NULL, pixelFormat, width, height, 0, &implData->texture[front]);
```
Comment (bghgary) [inline Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm:182]: nit
```suggestion
    CVReturn status = CVMetalTextureCacheCreateTextureFromImage(nullptr, implData->textureCache, pixelBuffer, nullptr, pixelFormat, width, height, 0, &implData->texture[front]);
```

File: Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm (hunk 4 above)
Comment (CedricGuillemet) [reply Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm:57]: With 2 textures, I still get shuttering sometimes. 3 is ok but I always favor power of 2. Which brings a question. For Apple and other platform, when user is updating a texture outside of BN, can it be a problem? For example, when a user updates a texture faster than rendering on another thread? Or can it happen in the middle of bgfx rendering?

**Comment by ryantrem** [comment]
The camera play return value part of this PR was fixed in this commit: https://github.com/BabylonJS/BabylonNative/commit/2d5c176a5e0a05bc5e02f0ae49ad9defe9a7df82

**Comment by CedricGuillemet** [comment]
Changes done in other PRs. Closing.

**Pull request closed without merging**